---
title:  "<big><u>IMAP-VISUALIZATION</u></big><br><br>Exploratory Analysis of 16S-Based Microbial Profiles<br><br>"
subtitle: |
  | ![](images/coders.png)<br>
  | Fostering Reproducible Microbiome data Analysis with Snakemake workflow
date:
- IMAP-VISUALIZATION<br>[Latest GitHub-Repo](https://github.com/tmbuza/imap-data-visualization/)<br>
- Maintained by Teresia Mrema-Buza<br><br>
- <b>Updated on `r Sys.Date()`</b>
author: Teresia Mrema-Buza
output: 
  html_document: 
css:
- styles.css
csl: "library/apa.csl"
bibliography:
- library/references.bib
- library/imap.bib
link-citations: true
mainfont:
biblio-style: apalike
email_address: "tmbuza@complexdatainsights.com"
github_repo: "https://github.com/tmbuza/imap-data-visualization"
description: |
  | Systematic microbiome data analysis with Snakemake workflows.
---

# Microbiome Data Visualization

<div id="main-container">

<link rel="preconnect" href="https://fonts.googleapis.com">
<link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
<link href="https://fonts.googleapis.com/css2?family=Anton" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Roboto:wght@100;300;400;500;700,900&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Oswald:wght@300;400;700&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Merriweather:wght@300;400;700&display=swap" rel="stylesheet">
<link href="https://fonts.googleapis.com/css2?family=Montserrat:wght@100;200;300;400;700&display=swap" rel="stylesheet">


```{css}
body h1 { margin-top: 1.8em; font-size: 2.0em; }
body h2 { margin-top: 1.8em; font-size: 2.0em; }
body h3 { margin-top: 1.8em; font-size: 1.6em; }
body h4 { margin-top: 1.8em; font-size: 1.2em; }
body h5 { margin-top: 1.8em; font-size: 1.0em; }
```

## Set global parameters

```{r globalsetup, echo=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(
  echo  =FALSE,
  message  =FALSE,
  warning  =FALSE,
  cache  =FALSE,
  comment  =NA,
  fig.path='./figures/',
  fig.show='asis',
  dev  ='svg',
  fig.align='center',
  out.width  ="70%",
  fig.width  =7,
  fig.asp  =0.7,
  fig.show  ="asis"
)
```

## Import libraries
```{r libraries}
if(!require("tidyverse")) {BiocManager::install("tidyverse")}
suppressPackageStartupMessages(library(tidyverse))
if(!require("schtools")) {BiocManager::install("schtools")}
library(schtools)
if(!require("tools")) {BiocManager::install("tools")}
library(tools)
if(!require("yaml")) {BiocManager::install("yaml")}
library(yaml)
if(!require("rhdf5")) {BiocManager::install("rhdf5")}
library(rhdf5)
if(!require("Matrix")) {BiocManager::install("Matrix")}
library(Matrix)
if(!require("qiime2R")) {devtools::install_github("jbisanz/qiime2R")}
library(qiime2R)
if(!require("microViz")) {install.packages("microViz",repos = c(davidbarnett = "https://david-barnett.r-universe.dev", getOption("repos")))}
library(microViz)
if(!require("funModeling")){devtools::install_github("pablo14/funModeling")}
suppressPackageStartupMessages(library("funModeling"))
if(!require("vegan")) {BiocManager::install("vegan")}
library(vegan)
if(!require("phyloseq")) {BiocManager::install("phyloseq")}
library(phyloseq)
if(!require("microbiome")) {BiocManager::install("microbiome")}
library(microbiome)
if(!require("ComplexHeatmap")) {BiocManager::install("ComplexHeatmap")}
library(ComplexHeatmap)


# load("data/processed_objects.rda", verbose=TRUE)
```





## Import test dataset
Using `dietswap` from `microbiome` package
```{r data}
library(microbiome)
data("dietswap", package = "microbiome")
ps <- dietswap
```


## Explore variables and missing values
```{r status}
library(microbiome)
library(phyloseq)
data("dietswap", package = "microbiome")

df <-dietswap %>% 
  phyloseq::psmelt() %>% 
  select(-sample) %>% 
  tibble::rownames_to_column("sample_id") %>% 
  rename_all(tolower)

library(funModeling)
df_status(df, print_results = FALSE) %>% 
  select(variable, type, unique, p_zeros) %>% tibble::tibble()
```


## Frequency distributions for categoric variables
```{r freq_catvars}
library(funModeling)
freq(df, 
     input = c("sex", "nationality", "group", "bmi_group"), 
     plot = TRUE,
     na.rm = FALSE)
```

## Correlation of numeric varaiables
```{r corr}
library(funModeling)
correlation_table(df, "sex")
```


## Boxplot
```{r boxplot}
library(funModeling)
plotar(df, input = c("sex"), target="bmi_group", plot_type="boxplot")
```

## Density histogram
```{r histdens}
library(funModeling)
plotar(df, input = c("sex", "nationality"), target="sex", plot_type="histdens")
```

# (PART) MACHINE LEARNING {-}
# Measuring model performance

![](images/gain_vs_roc.svg){width=100%}

## Cumulative Gains Charts with `funModeling`

Dataset: Heart disease dataset from the `funModeling` R package

### Load dataset and examine its structure
```{r}
library(funModeling)
data(heart_disease, package = "funModeling")
df <- heart_disease

df_status(df, print_results = FALSE) %>% 
  select(variable, type, unique, p_zeros) %>% tibble::tibble()
```

### Create machine learning model
```{r glm_model}
library(funModeling)

fit_glm=glm(has_heart_disease ~ age + oldpeak, data=df, family = binomial)
```


### Get the scores for positive case
```{r scores}
df$score=predict(fit_glm, newdata=df, type='response')
```

### Calculate performance metrics
```{r perfometrics, message=FALSE, warning=FALSE}
gain_lift(data=df, score='score', target='has_heart_disease')
```

# Coordinate plot: Profile mean clusters
```{r coord}
coord_plot(data=mtcars, group_var="cyl", group_func=median, print_table=TRUE)
```


# Visualization Snakemake workflow


![](dags/rulegraph.svg)


# Heatmaps


## Using `qiime2R`
```{r taxahtmp, eval=FALSE, include=FALSE}
metadata<-read_q2metadata("data/sample_metadata.tsv")
ASVs<-read_qza("data/feature_table.qza")$data
taxonomy<-read_qza("data/taxonomy.qza")$data %>% parse_taxonomy()

taxasums<-summarize_taxa(ASVs, taxonomy)$Genus

taxa_heatmap(taxasums, metadata, "BodySite")
```

## Using `ggplot`

Bray-Curtis
```{r bray}
ps <- ps

otutable <- otu_table(ps) %>% 
  psmelt() %>% 
  group_by(Sample) %>%
  mutate(N = sum(Abundance)) %>%
  ungroup()

n=min(otutable$N)
      
otutable <- otutable %>% 
  filter(N >= n) %>%
  select(-N) %>% 
  pivot_wider(names_from="OTU", values_from="Abundance", values_fill=0) %>%
  column_to_rownames("Sample")

## Getting Bray-`Curtis` distances
bray <- avgdist(otutable, dmethod="bray", sample=1776) %>%
  as.matrix() %>%
  as_tibble(rownames = "A") %>%
  pivot_longer(-A, names_to="B", values_to="distances")

bray %>%
  ggplot(aes(x=A, y=B, fill=distances)) +
  geom_tile() +
  theme(axis.text = element_blank()) +
  scale_fill_gradient(low="#FF0000", high="#FFFFFF", name=NULL)
```

Jaccard
```{r jaccard}
## Getting `Jaccard` distances
jaccard <- avgdist(otutable, dmethod="jaccard", sample=1776) %>%
  as.matrix() %>%
  as_tibble(rownames = "A") %>%
  pivot_longer(-A, names_to="B", values_to="distances")

jaccard %>%
  ggplot(aes(x=A, y=B, fill=distances)) +
  geom_tile() +
  theme(axis.text = element_blank()) +
  scale_fill_gradient(low="#FF0000", high="#FFFFFF", name=NULL)
```


Bray-Curtis and Jaccard
```{r bray_jcard}
labels <- tibble(
  x=c(50, 190),
  y=c(190, 30),
  label=c("Bray-Curtis", "Jaccard")
)

inner_join(bray, jaccard, by=c("A", "B")) %>%
  select(A, B, bray=distances.x, jaccard=distances.y) %>%
  mutate(distances = if_else(A < B, bray, jaccard)) %>%
  ggplot(aes(x=A, y=B, fill=distances)) +
  geom_tile() +
  geom_text(data=labels, aes(x=(x), y=y, label=label), inherit.aes=FALSE,
            size=10) +
  scale_fill_gradient(low="#FF0000", high="#FFFFFF", name=NULL) +
  labs(x="", y="") +
  theme_classic() +
  theme(axis.text = element_blank())
```


## Using `microViz`

Heatmap without sample annotation
```{r microvizhtmp}
library(tidyverse)
library(microViz)
library(phyloseq)
library(microbiome)

data("dietswap", package = "microbiome")
ps <- dietswap

ps %>%
  ps_mutate(nationality = as.character(nationality)) %>%
  tax_transform("log2", add = 1, chain = TRUE) %>%
  comp_heatmap(
    taxa = tax_top(ps, n = 30), grid_col = NA, name = "Log2p",
    taxon_renamer = function(x) stringr::str_remove(x, " [ae]t rel."),
    colors = heat_palette(palette = viridis::turbo(11)),
    row_names_side = "left", row_dend_side = "right", sample_side = "bottom",
  )

```

Heatmap with sample annotation
```{r micrvizannothtmp}
library(tidyverse)
library(microViz)
library(phyloseq)
library(microbiome)

data("dietswap", package = "microbiome")
ps <- dietswap

htmp <- ps %>%
  ps_mutate(nationality = as.character(nationality)) %>%
  tax_transform("log2", add = 1, chain = TRUE) %>%
  comp_heatmap(
    taxa = tax_top(ps, n = 30), grid_col = NA, name = "Log2p",
    taxon_renamer = function(x) stringr::str_remove(x, " [ae]t rel."),
    colors = heat_palette(palette = viridis::turbo(11)),
    row_names_side = "left", row_dend_side = "right", sample_side = "bottom",
    sample_anno = sampleAnnotation(
    Nationality = anno_sample_cat(
    var = "nationality", col = c(AAM = "red", AFR = "green"),
    box_col = NA, legend_title = "Nationality", size = grid::unit(4, "mm")
    )
  )
)
ComplexHeatmap::draw(
  object = htmp, annotation_legend_list = attr(htmp, "AnnoLegends"),
  merge_legends = TRUE
)

```

```{r}
library(tidyverse)
library(microViz)
library(phyloseq)
library(microbiome)

data("dietswap", package = "microbiome")
ps <- dietswap

ps %>%
  ps_mutate(nationality = as.character(nationality)) %>%
  tax_transform("log2", add = 1, chain = TRUE) %>%
  comp_heatmap(
    taxa = tax_top(ps, n = 30), grid_col = NA, name = "Log2p",
    taxon_renamer = function(x) stringr::str_remove(x, " [ae]t rel."),
    colors = heat_palette(palette = viridis::turbo(11)),
    row_names_side = "left", row_dend_side = "right", sample_side = "bottom",
  )

ggsave("figures/microviz_wo_heatmap.png", height=4, width=8, device="png")
ggsave("figures/microviz_wo_heatmap.svg", height=4, width=8, device="svg")


library(tidyverse)
library(microViz)
library(phyloseq)
library(microbiome)

data("dietswap", package = "microbiome")
ps <- dietswap

htmp <- ps %>%
  ps_mutate(nationality = as.character(nationality)) %>%
  tax_transform("log2", add = 1, chain = TRUE) %>%
  comp_heatmap(
    taxa = tax_top(ps, n = 30), grid_col = NA, name = "Log2p",
    taxon_renamer = function(x) stringr::str_remove(x, " [ae]t rel."),
    colors = heat_palette(palette = viridis::turbo(11)),
    row_names_side = "left", row_dend_side = "right", sample_side = "bottom",
    sample_anno = sampleAnnotation(
    Nationality = anno_sample_cat(
    var = "nationality", col = c(AAM = "red", AFR = "green"),
    box_col = NA, legend_title = "Nationality", size = grid::unit(4, "mm")
    )
  )
)
ComplexHeatmap::draw(
  object = htmp, annotation_legend_list = attr(htmp, "AnnoLegends"),
  merge_legends = TRUE
)

ggsave("figures/microviz_w_heatmap.png", height=4, width=8, device="png")
ggsave("figures/microviz_w_heatmap.svg", height=4, width=8, device="svg")

```

# Jitter plots
![](figures/q2r_jitterplot.svg)


# Line plots
![](figures/q2r_lineplot.svg)



# PCoA ordination
![](figures/q2r_pcoa.svg)

<br><hr width=100%><br>

# Related work
```{block imaprepos, echo=TRUE}


| Repo | Description | Status |
|:-----------------------|----------------------------------|--------------:|
| [IMAP-GLIMPSE](https://github.com/datainsights/imap-project-overview/) | IMAP project overview | [In-progress](https://datainsights.github.io/imap-project-overview/) |
| [IMAP-PART 01](https://github.com/tmbuza/imap-software-requirements/) | Software requirement for microbiome data analysis with Snakemake workflows | [In-progress](https://tmbuza.github.io/imap-software-requirements/) |
| [IMAP-PART 02](https://github.com/tmbuza/imap-sample-metadata/) | Downloading and exploring microbiome sample metadata from SRA Database | [In-progress](https://tmbuza.github.io/imap-sample-metadata/) |
| [IMAP-PART 03](https://github.com/tmbuza/imap-download-sra-reads/) | Downloading and filtering microbiome sequencing data from SRA database | [In-progress](https://tmbuza.github.io/imap-download-sra-reads/) |
| [IMAP-PART 04](https://github.com/tmbuza/imap-read-quality-control/) | Quality Control of Microbiome Next Generation Sequencing Reads | [In-progress](https://tmbuza.github.io/imap-read-quality-control/) |
| [IMAP-PART 05](https://github.com/tmbuza/imap-bioinformatics-mothur/) | Microbial profiling using MOTHUR and Snakemake workflows | [In-progress](https://tmbuza.github.io/imap-bioinformatics-mothur/) |
| [IMAP-PART 06](https://github.com/tmbuza/imap-bioinformatics-qiime2/) | Microbial profiling using QIIME2 and Snakemake workflows | [In-progress](https://tmbuza.github.io/imap-bioinformatics-qiime2/) |
| [IMAP-PART 07](https://github.com/tmbuza/imap-data-processing/) | Processing Output from 16S-Based microbiome bioinformatics pipelines| [In-progress](https://tmbuza.github.io/imap-data-processing/) |
| [IMAP-PART 08](https://github.com/tmbuza/imap-exploratory-analysis/) | Exploratory Analysis of 16S-Based Microbiome Processed Data | [In-progress](https://tmbuza.github.io/imap-exploratory-analysis/) |
| [IMAP-PART 09](https://github.com/tmbuza/imap-data-visualization/) | Visualization of Microbiome Data with the integration of Snakemake Workflows | [In-progress](https://visuals.complexdatainsights.com/) |
| [IMAP-PART XX](https://github.com/tmbuza/imap-machine-learning/) | Predictive Modeling of Microbiome Data Using Machine Learning Techniques | [In-progress](https://tmbuza.github.io/imap-machine-learning/) |
  [IMAP-SUMMARY](https://github.com/tmbuza/imap-snakemake-workflows/) | Summary of snakemake workflows for microbiome data analysis | [In-progress](https://imap.complexdatainsights.com/) |
```

## Citation

> Please consider citing the iMAP article [@TMBuza2019] if you find any part of the IMAP practical user guides helpful in your microbiome data analysis.

<br>

# References
::: {#refs}
:::

<br><hr width=100%><br>

# Appendix {-}

## Troubleshooting of FAQs
<ol>
  <li>Question</li>
    <ul>Answer</ul>
  <li>Question</li>
    <ul>Answer</ul>
</ol>


</div> <!-- end main-container -->

<br><hr width=100%><br>

<div id="footer">
Last updated on `r format(Sys.time(), '%B %d, %Y')`.<br>
Github_Repo: `r rmarkdown::metadata$github_repo`.<br>
Repo compiled and maintained by: [`r rmarkdown::metadata$author`](mailto:`r rmarkdown::metadata$email_address`). 
</div>

